src/Inventory/Asset/MainAsset.php

```diff
@@ -618,6 +618,23 @@ public function rulepassed($items_id, $itemtype, $rules_id, $ports_id = 0)
             }
         }
 
+        //check for any old agent to remove
+        $iterator = $DB->request([
+            'SELECT' => 'id',
+            'FROM' => $this->agent->getTable(),
+            'WHERE' => [
+                'itemtype' => $this->item->getType(),
+                'items_id' => $items_id,
+                'NOT' => [
+                    'id' => $this->agent->fields['id']
+                ]
+            ]
+        ]);
+        foreach ($iterator as $old_agent) {
+            $agent = new \Agent();
+            $agent->delete(['id' => $old_agent['id']], true);
+        }
```

Comment:
This should do the same.
```suggestion
        $agent = new \Agent();
        $agent->deleteByCriteria([
            'itemtype' => $this->item->getType(),
            'items_id' => $items_id,
            'NOT' => [
                'id' => $this->agent->fields['id']
            ]
        ]);
```